File: cpp/fileio.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <cstdlib>
#include <vector>

int main(int argc, char* argv[]) {
    const int fileCount = std::atoi(argv[1]);
    const int n = std::atoi(argv[2]);

    // Pre-allocate buffer with n '1' characters
    std::vector<char> write_buffer(n, '1');

    // Write phase - write entire buffer at once
    for (int i = 0; i < fileCount; ++i) {
        std::string filename = "file" + std::to_string(i);
        std::ofstream file(filename, std::ios::binary);
        file.write(write_buffer.data(), n);
        file.close();
    }

    // Read phase - read in chunks and count total bytes
    long long total_bytes = 0;
    const size_t READ_BUFFER_SIZE = 65536; // 64 KiB read buffer
    std::vector<char> read_buffer(READ_BUFFER_SIZE);
    
    for (int i = 0; i < fileCount; ++i) {
        std::string filename = "file" + std::to_string(i);
        std::ifstream file(filename, std::ios::binary);
        
        size_t bytes_read;
        while (file.read(read_buffer.data(), READ_BUFFER_SIZE) || 
               (bytes_read = file.gcount()) > 0) {
            if (file.gcount() > 0) {
                total_bytes += file.gcount();
            }
        }
        file.close();
    }

    // Count chunks of 65536 bytes
    const long long CHUNK_SIZE = 65536;
    const long long total_chunks = total_bytes / CHUNK_SIZE;
    std::cout << total_chunks << '\n';

    // Cleanup
    for (int i = 0; i < fileCount; ++i) {
        std::string filename = "file" + std::to_string(i);
        std::remove(filename.c_str());
    }

    return 0;
}

File: ocaml/fileio.ml
open Printf

let main () =
  let fileCount = int_of_string Sys.argv.(1) in
  let n = int_of_string Sys.argv.(2) in
  
  (* 1. Фаза записи *)
  
  (* Pre-allocate buffer with n '1' characters *)
  let write_buffer = String.make n '1' in
  
  (* Write phase - write entire buffer at once *)
  for i = 0 to fileCount - 1 do
    let filename = "file" ^ string_of_int i in
    let oc = open_out_bin filename in
    output_string oc write_buffer;
    close_out oc
  done;
  
  (* 2. Фаза чтения *)
  
  let total_bytes = ref 0 in
  let read_buffer_size = 65536 in  (* 64 KiB read buffer *)
  let read_buffer = Bytes.create read_buffer_size in
  
  (* Read phase - read in chunks and count total bytes *)
  for i = 0 to fileCount - 1 do
    let filename = "file" ^ string_of_int i in
    let ic = open_in_bin filename in
    
    let rec read_loop () =
      let bytes_read = input ic read_buffer 0 read_buffer_size in
      if bytes_read > 0 then (
        total_bytes := !total_bytes + bytes_read;
        read_loop ()
      )
    in
    
    read_loop ();
    close_in ic
  done;
  
  (* 3. Подсчет чанков и вывод результата *)
  
  let chunk_size = 65536 in
  let total_chunks = !total_bytes / chunk_size in
  printf "%d\n" total_chunks;
  
  (* 4. Очистка файлов *)
  
  for i = 0 to fileCount - 1 do
    let filename = "file" ^ string_of_int i in
    Sys.remove filename
  done

let () = main ()